### PR TITLE
feat: link implementation

### PR DESCRIPTION
**Pull request by alanpoulain**

### CODE REVIEW

File: docs/adr/0004-link.md [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+```php
+<?php
+
+#[Query]
+#[Get('/companies/{companyId}/employees', uriVariables: [
+    'companyId' => new Link(
+        targetClass: Company::class,
+        identifiers: ['id'],
+        property: 'company'
```
Comment (soyuka) [inline docs/adr/0004-link.md:34]: inverseProperty: 'employees' ?

File: src/Bridge/Doctrine/Orm/State/LinksHandlerTrait.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+                    }
+
+                    return;
+                }
+            }
+
+            throw new RuntimeException(sprintf('The class "%s" cannot be retrieved from "%s".', $resourceClass, $linkClass));
+        }
+
+        if (null === $links) {
```
Comment (soyuka) [inline src/Bridge/Doctrine/Orm/State/LinksHandlerTrait.php:61]:
```suggestion
        if (!$links) {
```

File: src/Metadata/Resource/Factory/UriTemplateResourceMetadataCollectionFactory.php [hunk 3]
```diff
@@ -178,150 +172,62 @@ private function normalizeUriVariables($operation)
...
                     if (2 !== \count($normalizedUriVariable)) {
                         throw new \LogicException("The uriVariables shortcut syntax needs to be the tuple: 'uriVariable' => [targetClass, targetProperty]");
                     }
-                    $normalizedUriVariable = (new UriVariable())->withIdentifiers([$normalizedUriVariable[1]])->withTargetClass($normalizedUriVariable[0]);
+                    $normalizedUriVariable = (new Link())->withIdentifiers([$normalizedUriVariable[1]])->withTargetClass($normalizedUriVariable[0]);
                 } else {
-                    $normalizedUriVariable = new UriVariable(null, $normalizedUriVariable['inverse_property'] ?? null, $normalizedUriVariable['property'] ?? null, $normalizedUriVariable['class'], $normalizedUriVariable['identifiers'] ?? null, $normalizedUriVariable['composite_identifier'] ?? null);
+                    $normalizedUriVariable = new Link(null, $normalizedUriVariable['inverse_property'] ?? null, $normalizedUriVariable['property'] ?? null, $normalizedUriVariable['class'], $normalizedUriVariable['identifiers'] ?? null, $normalizedUriVariable['composite_identifier'] ?? null);
```
Comment (soyuka) [inline src/Metadata/Resource/Factory/UriTemplateResourceMetadataCollectionFactory.php:184]: missing expanded_value

File: docs/adr/0004-link.md (hunk 1 above)
Comment (alanpoulain) [reply docs/adr/0004-link.md:34]: The `company` property is in the `Employee` class, it seems alright to me, isn't it?

File: docs/adr/0004-link.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+<?php
+
+#[Query]
+#[Get]
+class Company
+{
+  public $id;
+
+  /** @var Employee[] */
+  #[Link(inverseProperty: 'company')]
```
Comment (soyuka) [inline docs/adr/0004-link.md:55]:
```suggestion
  #[Link('company')]
```

File: docs/adr/0004-link.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+  public iterable $employees;
+}
+```
+
+```graphql
+{
+  companies(id: "/companies/2") {
+    employees {
+      edges {
+        node {
```
Comment (soyuka) [inline docs/adr/0004-link.md:65]: /companies/{id}/employees

Link id, targetClass: Company::class, property: 'company'

property => sur la classe du noeud qui suit le Link
targetProperty => sur la target class

**Comment by soyuka** [comment]
cherry picked in https://github.com/api-platform/core/pull/4537

**Pull request closed without merging**